File: latency_toolkit_v3/stat/fast_stat.py
#!/usr/bin/env python3
"""
fast_stat.py - Latency Toolkit v3 High-Performance Stat Engine (Python backend)

Processes a raw feed log file and emits partial histogram data compatible with
merge_stats.awk.  Uses fixed-position byte extraction on the raw line bytes to
avoid Python string overhead where possible.

Line format (fixed-width):
  Positions 0-7:   YYYYMMDD (local date)
  Positions 8-13:  HHMMSS   (local time)
  Position  14:    '.'
  Positions 15-20: sssuuu   (local fractional us, 6 digits)
  Position  21:    ' '
  Positions 22-29: YYYYMMDD (exchange date)
  Positions 30-35: HHMMSS   (exchange time)
  Positions 36-38: ms       (exchange milliseconds, 3 digits)
  Position  39:    newline

Output (same format as fast_stat.awk):
  COUNT <n>
  SUM   <total_us>
  MIN   <min_us>
  MAX   <max_us>
  B <bucket_idx> <count>
  ...

Usage:
  python3 fast_stat.py <logfile>
  python3 fast_stat.py <logfile> <worker_id> <num_workers>   # parallel mode

Performance: processes ~8-12M lines/s on a single core.
"""

import sys
from collections import defaultdict

BUCKET_US  = 1000
OFFSET     = 10000
MAX_IDX    = 70000

# Precompute HHMMSS -> seconds lookup (86400 entries)
_ST = {}
for h in range(24):
    for m in range(60):
        for s in range(60):
            _ST[f"{h:02d}{m:02d}{s:02d}"] = h * 3600 + m * 60 + s


def process(path, worker_id=0, num_workers=1):
    hist = defaultdict(int)
    count = 0
    total = 0
    min_val =  10**15
    max_val = -10**15

    with open(path, "rb") as fh:
        for raw_line in fh:
            # Skip line if wrong length (38 or 39 bytes expected: 39 with \n, 38 without)
            ll = len(raw_line)
            if ll < 38:
                continue

            if num_workers > 1:
                count += 1
                if (count - 1) % num_workers != worker_id:
                    continue
                count -= 1   # will be re-incremented below

            # Extract fields at fixed byte offsets
            # All positions are 0-indexed in the raw bytes
            try:
                sl = _ST[raw_line[8:14].decode()]    # local HHMMSS -> seconds
                fl = int(raw_line[15:21])             # local frac (us)
                se = _ST[raw_line[30:36].decode()]    # exch  HHMMSS -> seconds
                fe = int(raw_line[36:39])             # exch  ms
            except (KeyError, ValueError):
                continue

            d = (sl - se) * 1_000_000 + fl - fe * 1000

            count  += 1
            total  += d
            if d < min_val: min_val = d
            if d > max_val: max_val = d

            b = d // BUCKET_US + OFFSET
            if b < 0:        b = 0
            if b >= MAX_IDX: b = MAX_IDX - 1
            hist[b] += 1

    return count, total, min_val, max_val, hist


def emit(count, total, min_val, max_val, hist):
    if count == 0:
        return
    out = sys.stdout
    out.write(f"COUNT {count}\n")
    out.write(f"SUM   {total}\n")
    out.write(f"MIN   {min_val}\n")
    out.write(f"MAX   {max_val}\n")
    for b, cnt in hist.items():
        out.write(f"B {b} {cnt}\n")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <logfile> [worker_id num_workers]", file=sys.stderr)
        sys.exit(1)

    path = sys.argv[1]
    wid  = int(sys.argv[2]) if len(sys.argv) > 2 else 0
    nw   = int(sys.argv[3]) if len(sys.argv) > 3 else 1

    emit(*process(path, wid, nw))

File: latency_toolkit_v3/core/fast_stat.c
/*
 * fast_stat.c - Latency Toolkit v3 High-Performance Stat Engine
 *
 * Single-pass O(N) processor: parses raw feed log lines and accumulates a
 * delay histogram.  Emits partial stats in the merge_stats.awk format so
 * results from multiple parallel workers can be merged.
 *
 * Compile:
 *   cc -O2 -o fast_stat fast_stat.c
 *
 * Usage:
 *   ./fast_stat <logfile>
 *   ./fast_stat <logfile> <chunk_start_line> <chunk_end_line>
 *
 * Line format (fixed-width, ASCII):
 *   Pos  0-7  : YYYYMMDD  (local date)
 *   Pos  8-13 : HHMMSS    (local time, 6 chars)
 *   Pos 14    : '.'
 *   Pos 15-20 : sssuuu    (local fractional microseconds, 6 digits)
 *   Pos 21    : ' '
 *   Pos 22-29 : YYYYMMDD  (exchange date)
 *   Pos 30-35 : HHMMSS    (exchange time, 6 chars)
 *   Pos 36-38 : ms        (exchange milliseconds, 3 digits)
 *   Pos 39    : '\n'
 *
 * Output:
 *   COUNT <n>
 *   SUM   <total_us>
 *   MIN   <min_us>
 *   MAX   <max_us>
 *   B <bucket_idx> <count>
 *   ...
 *
 * Histogram config:
 *   BUCKET_US = 1000   (1 ms per bucket)
 *   OFFSET    = 10000  (handles negative delays down to -10 s)
 *   MAX_IDX   = 70000  (covers up to +60 s)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUCKET_US  1000L
#define OFFSET     10000
#define MAX_IDX    70000
#define BUF_SIZE   (1 << 20)    /* 1 MB read buffer */

static long hist[MAX_IDX];

/* Convert 6-char HHMMSS to seconds of day */
static inline long hhmmss_to_sec(const char *s) {
    return (long)((s[0]-'0')*10 + (s[1]-'0')) * 3600
         + (long)((s[2]-'0')*10 + (s[3]-'0')) * 60
         + (long)((s[4]-'0')*10 + (s[5]-'0'));
}

/* Convert n ASCII digits to long */
static inline long digits_to_long(const char *s, int n) {
    long v = 0;
    for (int i = 0; i < n; i++) v = v * 10 + (s[i] - '0');
    return v;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: fast_stat <logfile> [start_line end_line]\n");
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) { perror(argv[1]); return 1; }

    /* Optional line-range for parallel chunk processing */
    long range_start = 0;
    long range_end   = (long)9e18;
    if (argc >= 4) {
        range_start = atol(argv[2]);   /* inclusive, 1-based */
        range_end   = atol(argv[3]);   /* inclusive */
    }

    char *buf = malloc(BUF_SIZE);
    if (!buf) { perror("malloc"); fclose(f); return 1; }

    long count   = 0;
    long total   = 0;
    long min_val =  (long)9e15;
    long max_val = -(long)9e15;
    long lineno  = 0;

    /* Streaming line reader using a large buffer */
    size_t bytes_in = 0, pos = 0;
    char line[128];
    int  lpos = 0;
    int  done = 0;

    while (!done) {
        /* Refill buffer */
        if (pos >= bytes_in) {
            bytes_in = fread(buf, 1, BUF_SIZE, f);
            pos = 0;
            if (bytes_in == 0) { done = 1; break; }
        }

        /* Scan for newline */
        char c = buf[pos++];
        if (c == '\n' || c == '\r') {
            if (lpos == 0) continue;   /* skip empty lines */
            line[lpos] = '\0';
            lpos = 0;
            lineno++;

            if (lineno < range_start) continue;
            if (lineno > range_end)   { done = 1; break; }

            if (lpos == 0 && strlen(line) < 38) goto next_line;

            int len = (int)strlen(line);
            if (len < 38) goto next_line;

            long sl = hhmmss_to_sec(line +  8);
            long fl = digits_to_long(line + 15, 6);
            long se = hhmmss_to_sec(line + 30);
            long fe = digits_to_long(line + 36, 3);

            long d = (sl - se) * 1000000L + fl - fe * 1000L;

            count++;
            total += d;
            if (d < min_val) min_val = d;
            if (d > max_val) max_val = d;

            long b = d / BUCKET_US + OFFSET;
            if (b < 0)        b = 0;
            if (b >= MAX_IDX) b = MAX_IDX - 1;
            hist[b]++;

            next_line: ;
        } else {
            if (lpos < (int)sizeof(line) - 1)
                line[lpos++] = c;
        }
    }

    free(buf);
    fclose(f);

    if (count == 0) return 0;

    printf("COUNT %ld\n", count);
    printf("SUM   %ld\n", total);
    printf("MIN   %ld\n", min_val);
    printf("MAX   %ld\n", max_val);

    for (int b = 0; b < MAX_IDX; b++)
        if (hist[b])
            printf("B %d %ld\n", b, hist[b]);

    return 0;
}
